contracts/src/core/libraries/v3/usage-authorization/UsageAuthorizationLib.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import {SafeERC20, IERC20} from "lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IUsageAuthorizationRegistry} from "src/core/interfaces/IUsageAuthorizationRegistry.sol";
+import {UsageAuthorizationTypes} from "src/core/libraries/v3/usage-authorization/UsageAuthorizationTypes.sol";
+import {UsageAuthorizationStorage} from "src/core/libraries/v3/usage-authorization/UsageAuthorizationStorage.sol";
+
+library UsageAuthorizationLib {
+    using SafeERC20 for IERC20;
+
+    event ReservationAdded(
+        uint64 indexed quorumId, address indexed account, UsageAuthorizationTypes.Reservation reservation
+    );
+
+    event ReservationIncreased(
+        uint64 indexed quorumId, address indexed account, UsageAuthorizationTypes.Reservation reservation
+    );
+
+    event ReservationDecreased(
+        uint64 indexed quorumId, address indexed account, UsageAuthorizationTypes.Reservation reservation
+    );
+
+    event DepositOnDemand(uint64 indexed quorumId, address indexed account, uint256 amount, address indexed payer);
+
+    function s() internal pure returns (UsageAuthorizationStorage.Layout storage) {
+        return UsageAuthorizationStorage.layout();
+    }
+
+    /// @notice Adds a reservation for a user in a quorum. Requires that any previous reservation has ended, and the start of the new reservation is in the future.
+    function addReservation(
+        uint64 quorumId,
+        address account,
+        UsageAuthorizationTypes.Reservation memory reservation,
+        uint64 schedulePeriod
+    ) internal {
+        checkReservation(quorumId, reservation, schedulePeriod);
+
+        if (reservation.startTimestamp < s().quorum[quorumId].user[account].reservation.endTimestamp) {
+            revert IUsageAuthorizationRegistry.ReservationStillActive(
+                s().quorum[quorumId].user[account].reservation.endTimestamp
+            );
+        }
+        if (reservation.startTimestamp < block.timestamp) {
```

Comment:
checked with patrick on slack that this check can be safely removed, and allows for more flexibility that a reservation can be immediately active